pkg/depend/graph.go

```diff
@@ -0,0 +1,190 @@
+	// Read and parse the data.
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return NewGraph(data)
+}
+
+// NewGraph reads in the JSON data and creates the graph.
+func NewGraph(data []byte) (*Root, error) {
+	var reps []*repo
+	if err := json.Unmarshal(data, &reps); err != nil {
+		return nil, err
+	}
+	// Collect all the entries and check for duplications.
+	repos := make(map[string]*repo)
+	for _, repo := range reps {
+		// Populate the table
+		if _, ok := repos[repo.Name]; !ok {
+			repos[repo.Name] = repo
+		} else {
+			return nil, errors.New("Duplicate key: " + repo.Name)
+		}
+	}
+	return newGraphFromRepos(repos)
+}
+
+// GetNode will return the node and a bool for if it exists.
+func (root *Root) GetNode(name string) (*Node, bool) {
+	node, ok := root.Table[name]
+	return node, ok
+}
+
+// GetChildren returns a flat list of all children.
+// Does not contain duplicates.
+func (node *Node) GetChildren() []*Node {
+	visited := utils.NewSet()
+	var children []*Node
+	for _, d := range node.Deps {
+		if !visited.Exists(d.Name) {
+			visited.Add(d.Name)
+			children = append(children, d)
+		}
+		for _, c := range d.GetChildren() {
+			if !visited.Exists(c.Name) {
+				visited.Add(c.Name)
+				children = append(children, c)
+			}
+		}
+	}
+	return children
+}
+
+func newGraphFromRepos(repos map[string]*repo) (*Root, error) {
+	root := &Root{
+		Table:      make(map[string]*Node, len(repos)),
+		reposTable: repos,
+	}
+
+	_, err := root.topologicalSort()
+	if err != nil {
+		return nil, err
+	}
+	root.createGraph()
+	return root, nil
+}
+
+// Uses depth first
+func (root *Root) topologicalSort() ([]*Node, error) {
+	//TODO: Refactor this...
+	return root.resolve(root.reposTable, nil)
+}
+
+// resolveFromRepos will take the parsed JSON and resolve it into the Root.Table
+func (root *Root) resolveFromRepos(repos map[string]*repo, visited *utils.StringSet) ([]*Node, error) {
+	for k, v := range repos {
+		for _, d := range v.Deps {
+			if visited.Exists(d) {
+				return nil, &NodeCycleError{k, d}
+			}
+		}
+	}
+	return root.resolve(repos, visited)
+}
+
+func (root *Root) resolve(repos map[string]*repo, visited *utils.StringSet) ([]*Node, error) {
+	var top bool
+	if visited == nil {
+		top = true
+	}
+
+	var nodes []*Node
+	for k, v := range repos {
+		if top {
+			visited = utils.NewSet()
+		}
+		visited.Add(k)
+		node, ok := root.Table[k]
+		// Check to see if we have populated this node already.
+		if !ok {
+			deps := make(map[string]*repo)
+			for _, d := range v.Deps {
+				deps[d] = root.reposTable[d]
+			}
+			nod, err := root.resolveFromRepos(deps, visited)
+			if err != nil {
+				return nil, err
+			}
+			node = &Node{
+				Name: v.Name,
+				URL:  v.URL,
+				Deps: nod,
+			}
+			root.Table[k] = node
+			nodes = append(nodes, node)
+		} else {
+			nodes = append(nodes, node)
+		}
+	}
+	return nodes, nil
+}
+
+// createGraph will find the edges and populate the root.
+func (root *Root) createGraph() {
```

Comment:
It's not clear to me if `createGraph` is needed. Doesn't the recursive traversal occurring in `resolve` allow for identification of all deps (and detection of any cycles) and thus allow for the identification of all edges?

Reply:
done